src/persist/src/postgres.rs

```diff
@@ -158,16 +168,49 @@ impl PostgresConsensus {
             },
         );
 
+        let last_ttl_connection = AtomicU64::new(0);
+        let connections_created = config.metrics.connpool_connections_created.clone();
+        let ttl_reconnections = config.metrics.connpool_ttl_reconnections.clone();
         let pool = Pool::builder(manager)
             .max_size(config.connection_pool_max_size)
-            .post_create(Hook::async_fn(|client, _| {
+            .post_create(Hook::async_fn(move |client, _| {
+                connections_created.inc();
                 Box::pin(async move {
                     debug!("opened new consensus postgres connection");
                     client.batch_execute(
                         "SET SESSION CHARACTERISTICS AS TRANSACTION ISOLATION LEVEL SERIALIZABLE",
                     ).await.map_err(|e| HookError::Abort(HookErrorCause::Backend(e)))
                 })
             }))
+            .pre_recycle(Hook::sync_fn(move |_client, conn_metrics| {
+                // proactively TTL connections to rebalance load to Postgres/CRDB. this helps
+                // fix skew when downstream DB operations (e.g. CRDB rolling restart) result
+                // in uneven load to each node, and works to reduce the # of connections
+                // maintained by the pool after bursty workloads.
+
+                // add a bias towards TTLing older connections first
+                if conn_metrics.age() < config.connection_pool_ttl {
+                    return Ok(());
+                }
+
+                let last_ttl = last_ttl_connection.load(Ordering::SeqCst);
+                let now = (SYSTEM_TIME)();
+                let elapsed_since_last_ttl = Duration::from_millis(now.saturating_sub(last_ttl));
+
+                // stagger out reconnections to avoid stampeding the DB
+                if elapsed_since_last_ttl > config.connection_pool_ttl_stagger
+                    && last_ttl_connection
+                        .compare_exchange_weak(last_ttl, now, Ordering::SeqCst, Ordering::Relaxed)
```

Comment:
I have a hard time reasoning about any ordering other than SeqCst and I don't really think the perf will be an issue here, so I'd say we should just do the most obvious thing (which I think is compare_exchange and SeqCst everywhere?). but that's not a hard preference if you're confidant this is correct as-is

Reply:
I thiiink we're OK since the second `Ordering` is about the failure case (`last_ttl` fails the comparison) and we toss away the return, but I'm def not confident enough in that to say for sure. SeqCst works for me :D